text/0000-decorators.md

```diff
@@ -0,0 +1,1212 @@
+receives the string `'prefix'`, so it has no context for how it will be used.
+
+The native class piece of this puzzle is completely inflexible. A decorator must
+be a function, there is no choice about it. However, the _Ember_ piece is _very_
+flexible. The legacy object model just needs a way to get the meta information
+for the property when the class is being finalized. We can either assign the
+meta information to the decorator function directly, or we can associate it via
+a `WeakMap`.
+
+The benefit of doing this is that the entire Ember ecosystem will get decorator
+support with no extra work required. Because standard `computed` definitions
+will work as decorators, existing macros will also work as decorators with no
+changes to existing code.
+
+All of Ember's built in computed macros in the `@ember/object/computed` module
+will also become decorators with no extra work. However, the injection macros
+will require slight updates as they use a subclass of the `ComputedProperty`
+class. These updates should be relatively minor, and will follow the same
+strategy as `computed()`.
+
+#### Usage and API
+
+The API for `computed` will remain mostly the same. The key differences will be:
+
+1. The result of `computed` will be a decorator which can be applied directly to
+   native _getters_, _setters_, and _class fields_.
+2. The `ComputedPropertyConfig` (the getter/setter functions) argument provided
+   to `computed` will now be optional when used as a decorator on a native
+   getter or setter, and the native getter/setter will be used instead.
+
+```ts
+function computed(...args: (string | ComputedPropertyConfig)[]): PropertyDecorator;
+```
+
+The function signatures of all existing macros, including `inject` macros, would
+change in the same way.
+
+In general usage, these three definitions are equivalent:
+
+```js
+import EmberObject, { computed } from '@ember/object';
+
+const Person = EmberObject.extend({
+  fullName: computed('firstName', 'lastName', {
+    get() {
+      return `${this.firstName} ${this.lastName}`;
+    }
+  })
+});
+
+class Person {
+  @computed('firstName', 'lastName')
+  get fullName() {
+    return `${this.firstName} ${this.lastName}`;
+  }
+}
+
+class Person {
+  @computed('firstName', 'lastName', {
+    get() {
+      return `${this.firstName} ${this.lastName}`;
+    }
+  })
+  fullName;
+}
+```
+
+That last example may seem confusing at first, but this is actually the same as
+defining a computed property macro:
+
+```js
+import EmberObject, { computed } from '@ember/object';
+
+function join(...dependentKeys) {
+  return computed(...dependentKeys, {
+    get() {
+      return dependentKeys.map(key => this[key]).join(' ');
+    }
+  });
+}
+
+class Person {
+  @join('firstName', 'lastName')
+  fullName;
+}
+```
+
+Notably, using computed as a decorator _directly_ will not be supported. This is
+to prevent a parameter check on a critical path:
+
+```js
+class Person {
+  @computed
```

Comment:
`@computed()` would also work, right?

Reply:
yes, absolutely. It shouldn't be necessary _most_ of the time, but it will work.

Reply:
Good for lazy computation with caching I suppose

Reply:
Right, that's a significant difference. People use this pattern not only for caching but also to make the reification process lazy. Whether it makes a noticeable performance impact is debatable, but I could see it mattering in some cases.

Reply:
I'll add a callout to that then, so people know it can still be used